README.md
# Azure Demo

This branch of the project is integrated with a *storage account queue* named `healthbeat`, putting a message each 15 seconds into as a heartbeat. See [pokemon-dashboard](https://github.com/ciberado/pokemon-dashboard) to visualize the result.

To deploy this application run the following script as *custom-data* or *custom-script extension* in any **Ubuntu** based vm.

PLEASE: update the script before using it replacing the connection string with credentials for accessing your queue.

```bash
#!/bin/sh

sudo apt update
sudo apt -y install curl dirmngr apt-transport-https lsb-release ca-certificates
curl -sL https://deb.nodesource.com/setup_12.x | sudo -E bash -
sudo apt -y install nodejs gcc g++ make

git clone https://github.com/ciberado/pokemon-nodejs
cd pokemon-nodejs
git checkout azure-demo
npm install
export PORT=8080
export AZURE_STORAGE_CONNECTION_STRING="DefaultEndpointsProtocol=<...>core.windows.net"
node app.js
```

See [SDK documentation](https://azuresdkdocs.blob.core.windows.net/$web/javascript/azure-storage-queue/12.0.0/index.html) for further information.

CHANGELOG.md
# Changelog

## 1.0.2

* Added 'DELETE /?now' to immediately exit the process with an error
* Type fixed

## 1.0.1

* Added proper changelog
* Removed unnecessary CSS
* Now correctly managing `Accept: application/json` header
* Showing the list of friends (from other servers)

app.js



const express = require('express');
const path = require('path');
const favicon = require('serve-favicon');
const logger = require('morgan');
const cookieParser = require('cookie-parser');
const bodyParser = require('body-parser');
const exphbs  = require('express-handlebars');

const routes = require('./routes/index');

const app = express();

const env = process.env.NODE_ENV || 'development';
app.locals.ENV = env;
app.locals.ENV_DEVELOPMENT = env == 'development';

let baseUrl = process.env.BASE_URL || '/';
if (baseUrl.endsWith('/') === true) {
   baseUrl = baseUrl.substring(1);
}
if (baseUrl.startsWith('/') === false) {
   baseUrl = '/' + baseUrl;
}

console.log(`Using ${baseUrl} as url base.`);

// view engine setup

app.engine('handlebars', exphbs({
  defaultLayout: 'main',
  partialsDir: ['views/partials/']
}));
app.set('views', path.join(__dirname, 'views'));
app.set('view engine', 'handlebars');

// app.use(favicon(__dirname + '/public/img/favicon.ico'));
app.use(logger('dev'));
app.use(bodyParser.json());
app.use(bodyParser.urlencoded({
  extended: true
}));
app.use(cookieParser());
app.use(baseUrl, express.static(path.join(__dirname, 'public')));

app.use(baseUrl, routes);

/// catch 404 and forward to error handler
app.use((req, res, next) => {
    const err = new Error('Not Found');
    err.status = 404;
    next(err);
});

/// error handlers

// development error handler
// will print stacktrace
if (app.get('env') === 'development') {
    app.use((err, req, res, next) => {
        res.status(err.status || 500);
        res.render('error', {
            message: err.message,
            error: err,
            title: 'error'
        });
    });
}

// production error handler
// no stacktraces leaked to user
app.use((err, req, res, next) => {
    res.status(err.status || 500);
    res.render('error', {
        message: err.message,
        error: {},
        title: 'error'
    });
});

module.exports = app;


pokemon.js
const { SecretClient } = require("@azure/keyvault-secrets");
const { DefaultAzureCredential } = require("@azure/identity");
const sql = require('mssql')
const delay = require('delay');

if (!process.env.KEY_VAULT_URI) {
  console.error('Please, define KEY_VAULT_URI environment variable. '+
    'For example: \nexport KEY_VAULT_URI=https://lab-vault.vault.azure.net');
  process.exit(-1);
}

async function getConnectionConfig() {
  let config;
  const credentials = new DefaultAzureCredential();
  const client = new SecretClient(process.env.KEY_VAULT_URI, credentials);
  const secret = await client.getSecret('PokemonDBConn');
  const c = secret.value;
  const connAttr = new Map(c.split(';').map(l => [l.split('=')[0].toLowerCase(), l.split('=')[1]]));

  config = {
    user: connAttr.get('uid'),
    password: connAttr.get('pwd'),
    server: connAttr.get('server').match('.*:(.*),')[1], 
    database: connAttr.get('database'),
    connectionTimeout : 1000*30,
    encrypt : true
  };  

  return config;
}

async function queryDatabaseForRandomPokemon(config) {
  let pool;
  try {
    const query="SELECT TOP 1 id, name FROM pokemon ORDER BY NEWID()";
    pool = await sql.connect(config);
    const result = await pool.request().query(query);
    const pokemon = result.recordset[0];

    return pokemon;
  } finally {
    if (pool) pool.close();
  }
}

async function getRandomPokemon() {
  let config;
  do {
    try {
      config = await getConnectionConfig();
    } catch (err) {
      console.error(`Error accessing Vault ${process.env.KEY_VAULT_URI}: ${err}. Retrying in 60 seconds.`);
      await delay(60*1000);
    }  
  } while (!config);
  let pokemon;
  do {
    try {
      pokemon = await queryDatabaseForRandomPokemon(config);
    } catch (err) {
      console.error(`Error accessing DB ${config.database}: ${err}. Retrying in 60 seconds.`);
      await delay(30*1000);
    }  
  } while (!pokemon);
  return pokemon;
}


module.exports = {
  getRandomPokemon : getRandomPokemon
};


healthbeat.js
const { DefaultAzureCredential } = require("@azure/identity");
const { QueueServiceClient } = require("@azure/storage-queue");
const rp = require('request-promise-native');
const delay = require('delay');

const QUEUE_NAME = 'healthbeats';
const INTERVAL_IN_SECONDS = 5;

let queueClient;
const info = {};

async function getNodeIPIfconfigCo() {
  const options = {
    uri: 'http://ifconfig.co/ip'
  };
  const publicIp = await rp(options);
  return publicIp;
}

async function getNodeIPMetadata() {
  // https://docs.microsoft.com/es-es/azure/virtual-machines/windows/instance-metadata-service
  const options = {
    uri: 'http://169.254.169.254/metadata/instance/network',
    qs : {
      'api-version' : '2017-08-01'
    },
    headers : {
        'Metadata' : 'true'
    },
    json : true
  };

  const result = await rp(options);
  const publicIp = result.interface[0].ipv4.ipAddress[0].publicIpAddress;
  return publicIp;
}

async function getNodeIP() {
  let publicIp;
  do {
    publicIp = await getNodeIPMetadata();
    if (!publicIp) {
      publicIp = await getNodeIPIfconfigCo();
      if (publicIp) publicIp = publicIp.trim();
    }
    if (!publicIp) {
      console.log(`Still waiting for public ip. Retrying in 20 seconds.`);
      await delay(20*1000);
    }
  } while(!publicIp);
  return publicIp;
}

async function emitHealthbeat(beat) {
  const msgAsString = JSON.stringify({beat});
  console.debug(`Healthbeat message: ${msgAsString}.`);
  return queueClient.sendMessage(msgAsString);
}


async function main() {
  const ip = await getNodeIP();
  console.log(`Starting healthbeat. Computer IP: ${ip}.`);
  if (!process.env.AZURE_STORAGE_NAME) {
    console.warn(`AZURE_STORAGE_NAME variable not found.`+ 
                 `Proceeding without healthbeat.`);
    return;
  }
 
  const defaultAzureCredential = new DefaultAzureCredential();
  const queueServiceClient = new QueueServiceClient(
    `https://${process.env.AZURE_STORAGE_NAME}.queue.core.windows.net`,
    // Example: `https://lab1018storageaccount.queue.core.windows.net`,
    defaultAzureCredential, {
      retryOptions: { 
        maxTries: 4 
      }
    }
  );
  

  do {
    try {
      queueClient = queueServiceClient.getQueueClient(QUEUE_NAME);  
    } catch(err) {
      console.error(`Error accessing queue ${QUEUE_NAME}: ${err}. Retrying in 60 seconds.`);
      await delay(60*1000);
    }
  } while (!queueClient);
    
  setInterval(async () => {
    try {
      await emitHealthbeat({ip, info, timestamp : Date.now()});
    } catch (err) {
      console.warn(`ERR: healthbet error (${err}).`);
    }
  }, INTERVAL_IN_SECONDS * 1000);
}

module.exports.info = info;
module.exports.start = main;

healthping.js
const azure = require('azure-storage');
const rp = require('request-promise-native');

const QUEUE_NAME = 'healthbeats';
const INTERVAL_IN_SECONDS = 15;

const healthbeatData = {}

if (!process.env.AZURE_STORAGE_CONNECTION_STRING) {
    console.warn(`AZURE_STORAGE_CONNECTION_STRING variable not found.`+ 
                 `Proceeding without healthbeat.`);
    return;
}


const queueSvc = azure.createQueueService();

async function getNodeIP() {
    const options = {
      uri: 'http://169.254.169.254/metadata/instance/network',
      qs : {
        'api-version' : '2017-08-01'
      },
      headers : {
          'Metadata' : 'true'
      },
      json : true
    };

    const result = await rp(options);
    return result.interface[0].ipv4.ipAddress[0].publicIpAddress;
}

async function emitHealthbeat(body) {
  healthbeatData.body = body;
  const msgAsString = JSON.stringify(healthbeatData);
  console.debug(`Healthbeat message: ${msgAsString}.`);
  return new Promise(function(resolve, reject) {
    queueSvc.createMessage(QUEUE_NAME, msgAsString, function(err, result, response) {
      if (err) reject(err)
      else resolve({result, response});
    })
  });
}


async function main() {
  const ip = await getNodeIP();
  console.log(`Computer IP: ${ip}.`);
  setInterval(() => emitHealthbeat({ip}), INTERVAL_IN_SECONDS * 1000);
}

main();

module.exports.healthbeatData = healthbeatData;

//https://docs.microsoft.com/en-us/azure/storage/queues/storage-nodejs-how-to-use-queues


src/index.js
const os =  require('os');
const process = require('process');

const fetch = require('node-fetch');

const express = require('express');
const favicon = require('serve-favicon');
const cookieParser = require('cookie-parser');
const bodyParser = require('body-parser');
const { engine } = require('express-handlebars');

const pokemonDB = require('./pokemon');

const pokemon = pokemonDB.getRandomPokemon();
console.log(`Selected pokemon: %s.`, JSON.stringify(pokemon));

const services = process.env.SERVICES?.split(',') || [];
console.log(`Services: ${JSON.stringify(services)}.`);

process.on('SIGTERM', (evt)=>{
  console.log(`Finishing process (${pokemon.name}).`);
  process.exit(0);
});

let healthy = true;

const app = express();

let baseUrl = process.env.BASE_URL || '/';
if (baseUrl.endsWith('/') === false) {
   baseUrl = baseUrl + '/';
}
if (baseUrl.startsWith('/') === false) {
   baseUrl = '/' + baseUrl;
}
console.log(`Hostname: ${process.env.HOSTNAME}.`)
console.log(`Using ${baseUrl} as url base.`);

app.get('/', async (req, res) => {
  const friends = [];
  
  await Promise.all(services.map(async s=>{
    console.log(`Invoking service ${s}`);
    const fetchResponse = await fetch(s, { 
      headers : {
        'Accept' : 'application/json'
      }
    });
    const fetchResponseJSON = await(fetchResponse.json());
    friends.push(fetchResponseJSON);
    console.log(`Service ${s} completed (${fetchResponseJSON.pokemon.name}).`);
  }));

  const resPayload = {
    pokemon,
    hostname : os.hostname(),
    friends : friends
  };

  if (req.get('Accept') === 'application/json') {
    console.log(`Rendering json`);
    res.setHeader('Content-Type', 'application/json');
    res.end(JSON.stringify(resPayload));
  } else {
    console.log(`Rendering webpage.`);
    res.render('index', { 
      title: 'Pokemon' , 
      payload : resPayload
    }), (err, html) => {
      res.end();
    };  
  }

});

app.get('/health', (req, res) => {
  if (healthy === false) {
    res.status(500).send('I really don\'t feel well.');
  } else {
    res.send('Looks everything is ok.');
  }
});

app.delete('/', (req, res) => {
  healthy = false;
  if (req.query.now !== undefined) {
    process.exit(-1);
  }
  res.send('Done');
});


app.engine('handlebars', engine({
  defaultLayout: 'main'
}));
app.set('views', 'views');
app.set('view engine', 'handlebars');

app.use(favicon('public/img/favicon.ico'));

app.use(bodyParser.json());
app.use(bodyParser.urlencoded({
  extended: true
}));
app.use(cookieParser());
app.use(express.static('public'));


app.use((req, res, next) => {
    const err = new Error('Not Found');
    err.status = 404;
    next(err);
});

app.use((err, req, res, next) => {
    res.status(err.status || 500);
    res.render('error', {
        message: err.message,
        error: {},
        title: 'error'
    });
});

app.set('port', process.env.PORT || 80);

var server = app.listen(app.get('port'), function() {
  console.log('Express server listening on port ' + server.address().port);
});
